### PR TITLE
Untangling setUnsafe()

### PR DESCRIPTION
**Pull request by WalterBright**
The setUnsafe() function is a horrible mess with multiple overloads and undocumented dependencies on various argument compinations. I'm going to try an untangle it.

### CODE REVIEW

**Comment by WalterBright** [comment]
This installment separates out the special case of function into a separate function, with a separate name instead of an overload.

**Review by thewilsonator — APPROVED** [review]

File: compiler/src/dmd/safe.d [hunk 1]
```diff
@@ -373,6 +374,22 @@ extern (D) bool setUnsafe(
...
 
+extern (D) bool setFunctionToUnsafe(FuncDeclaration fd)
+{
+    if (fd.safetyInprocess)
+    {
+        fd.safetyInprocess = false;
+        fd.type.toTypeFunction().trust = TRUST.system;
+
+        if (fd.fes)
+            setFunctionToUnsafe(fd.fes.func);
```
Comment (dkorpel) [inline compiler/src/dmd/safe.d:385]: I understand the renaming, but I don't think it's a good idea to duplicate this function with partially evaluated arguments: it only adds new confusing overloading, and every call to this function is basically a diagnostic bug and should be replaced with a call to the first overload of `setFunctionToUnsafe`, which explains to the user why a function is being marked unsafe.

Comment (WalterBright) [reply compiler/src/dmd/safe.d:385]: The problems are:

1. the multiple different overloadings of setUnsafe
2. the tangle of different behaviors taken based on whether fmt is null or arg is null or both are null
3. fhe absurdly complicated forwarding of the optional arguments

I've spent far too much time trying to trace through the logic of this function and its partner `errorSupplementalInferredAttr()`

My plan is to replace this rube goldberg machine with proper use of `va_list`. Problem (2) should split the function into 3 distinct functions. This particular PR factors out the null null case.

It's my fault that va_list wasn't used in the original incarnation of this, and I aim to fix it. The first two steps are done - this one, and improving ErrorSink to support va_list.

Comment (WalterBright) [reply compiler/src/dmd/safe.d:385]: The solution to the `arg ? toChars(arg) : null` inefficiency is to split the function into two:

1. determine if there is an error
2. if so, then produce the error message

AttributeViolation can use OutBuffer.printf to produce a string, as you are correct in that the `va_list` cannot be saved.

**Pull request merged**